DATACMNS-764 - Disambiguate custom repository implementation

schauder opened the pull request:
Disambiguates custom implementations of repositories by comparing the bean names of the candidate implementations with the bean name of the repository.

Therefore naming the unwanted implementations or naming the repository interface and the required implementation can be used to break ambiguities.

Also, fixes the path on which test sources get generated.

olivergierke commented on src/main/asciidoc/repositories.adoc:
```diff
@@ -649,6 +649,39 @@ If you use namespace configuration, the repository infrastructure tries to autod
+----
+package com.acme.impl.one;
+
+class UserRepositoryImpl implements UserRepositoryCustom {
```
I think we can even leave away the method declarations here. Should we describe how one would get the second implementation wired?

olivergierke commented on src/main/java/org/springframework/data/repository/config/AnnotationRepositoryConfigurationSource.java:
```diff
@@ -87,6 +92,8 @@ public AnnotationRepositoryConfigurationSource(AnnotationMetadata metadata, Clas
 		this.resourceLoader = resourceLoader;
 		this.hasExplicitFilters = hasExplicitFilters(attributes);
+
+		beanNameGenerator = new RepositoryBeanNameGenerator(resourceLoader.getClassLoader());
```
Qualify with `this.`.

olivergierke commented on src/main/java/org/springframework/data/repository/config/AnnotationRepositoryConfigurationSource.java:
```diff
@@ -46,6 +48,7 @@
  * @author Peter Rietzler
+ * @author Jens Schauder
  */
 public class AnnotationRepositoryConfigurationSource extends RepositoryConfigurationSourceSupport {
```
Couldn't all of this go into `RepositoryConfigurationSourceSupport`?

olivergierke commented on src/main/java/org/springframework/data/repository/config/CustomRepositoryImplementationDetector.java:
```diff
@@ -47,6 +50,7 @@
 	private final MetadataReaderFactory metadataReaderFactory;
 	private final Environment environment;
 	private final ResourceLoader resourceLoader;
+	private final RepositoryBeanNameGenerator nameGenerator;
```
I wonder if there's a way that we can avoid holding an additional generator to the one that the configuration uses. I could live with the CDI interaction becoming a bit more complicated for now as we're going to unify that into a dedicated configuration interface anyway and it would even make the requirements for such a unified interface even more explicit.

So let's say the `detectCustomImplementation(…)` overload taking a lot of parameters doesn't have any bounds for now. Let's make that as complicated as necessary to then know what needs to go into the shared config interface.

olivergierke commented on src/main/java/org/springframework/data/repository/config/CustomRepositoryImplementationDetector.java:
```diff
@@ -80,23 +86,55 @@ public AbstractBeanDefinition detectCustomImplementation(RepositoryConfiguration
 
+		Set<BeanDefinition> definitions = findCandidateBeanDefinitions(className, basePackages, excludeFilters);
+
+		if (definitions.isEmpty()) {
```
Couldn't that logic go into `findCandidateBeanDefinitions(…)` directly?

olivergierke commented on src/main/java/org/springframework/data/repository/config/RepositoryBeanNameGenerator.java:
```diff
@@ -15,56 +15,68 @@
 	 * @return
 	 */
-	private Class<?> getRepositoryInterfaceFrom(BeanDefinition beanDefinition) {
+	Class<?> getRepositoryInterfaceFrom(BeanDefinition beanDefinition) {
```
Can this stay private?

olivergierke commented on src/test/java/org/springframework/data/repository/config/DefaultRepositoryConfigurationUnitTests.java:
```diff
@@ -17,29 +17,44 @@
+	public void before() {
+
+		when(source.getBeanNameGenerator()).thenReturn(
+				new RepositoryBeanNameGenerator(Thread.currentThread().getContextClassLoader())
```
Any reason you're using this over `getClass().getClassLoader()`?

olivergierke commented on src/main/java/org/springframework/data/repository/config/CustomRepositoryImplementationDetector.java:
```diff
@@ -80,23 +84,43 @@ public AbstractBeanDefinition detectCustomImplementation(RepositoryConfiguration
+		Set<BeanDefinition> definitions = findCandidateBeanDefinitions(className, basePackages, excludeFilters);
+
+		SelectionSet<BeanDefinition> selection = new SelectionSet<>(definitions, c -> {
+			if (c.isEmpty()) {
```
Ternary if might allow us to condense this into a single line to avoid the block.

olivergierke commented on src/main/java/org/springframework/data/repository/config/CustomRepositoryImplementationDetector.java:
```diff
@@ -80,23 +84,43 @@ public AbstractBeanDefinition detectCustomImplementation(RepositoryConfiguration
+			}
+			return throwAmbiguousCustomImplementationException(c);
+		});
+		selection.filter(bd -> beanName != null && beanName.equals(nameGenerator.generateBeanName(bd)));
```
Blank line before (generally alway after a block unless it's only a closing `}`).

olivergierke commented on src/main/java/org/springframework/data/repository/config/CustomRepositoryImplementationDetector.java:
```diff
@@ -117,15 +141,10 @@ public AbstractBeanDefinition detectCustomImplementation(String className, Itera
 		for (String basePackage : basePackages) {
 			definitions.addAll(provider.findCandidateComponents(basePackage));
 		}
+		return definitions;
```
Blank line before (see above).

olivergierke commented on src/main/java/org/springframework/data/repository/config/CustomRepositoryImplementationDetector.java:
```diff
@@ -117,15 +141,10 @@ public AbstractBeanDefinition detectCustomImplementation(String className, Itera
-		}
-
+	private static AbstractBeanDefinition throwAmbiguousCustomImplementationException(Collection<BeanDefinition> definitions) {
 		List<String> implementationClassNames = new ArrayList<String>();
```
Blank line before multi-line method implementations.

olivergierke commented on src/main/java/org/springframework/data/repository/config/DefaultRepositoryConfiguration.java:
```diff
@@ -114,7 +115,7 @@ public String getImplementationClassName() {
 	 */
 	public String getImplementationBeanName() {
-		return StringUtils.uncapitalize(getImplementationClassName());
+		return configurationSource.getBeanNameGenerator().generateBeanName(definition)  + getImplementationPostfix();
```
Assign generator to local variable. Avoid double spaces before `+`.

olivergierke commented on src/main/java/org/springframework/data/repository/config/RepositoryBeanNameGenerator.java:
```diff
@@ -15,56 +15,68 @@
 	 * @return
 	 */
 	private Class<?> getRepositoryInterfaceFrom(BeanDefinition beanDefinition) {
+		if (beanDefinition instanceof ScannedGenericBeanDefinition) {
```
Blank line before multi-line method implementations.

olivergierke commented on src/main/java/org/springframework/data/repository/config/RepositoryBeanNameGenerator.java:
```diff
@@ -15,56 +15,68 @@
+			} catch (ClassNotFoundException e) {
+				throw new IllegalStateException("Could not resolve bean class.", e);
+			}
+		} else {
```
We don't need an else-block here as the first branch always yields a result, right?

olivergierke commented on src/main/java/org/springframework/data/repository/config/RepositoryBeanNameGenerator.java:
```diff
@@ -15,56 +15,68 @@
+	}
 
+	private Class<?> getRepositoryInterfaceFromFactory(BeanDefinition beanDefinition) {
 		Object value = beanDefinition.getConstructorArgumentValues().getArgumentValue(0, Class.class).getValue();
```
Blank line before multi-line method implementations.

olivergierke commented on src/main/java/org/springframework/data/repository/config/RepositoryConfigurationDelegate.java:
```diff
@@ -133,7 +128,7 @@ private static Environment defaultEnvironment(Environment environment, ResourceL
 
 			AbstractBeanDefinition beanDefinition = definitionBuilder.getBeanDefinition();
-			String beanName = beanNameGenerator.generateBeanName(beanDefinition, registry);
+			String beanName = configurationSource.getBeanNameGenerator().generateBeanName(beanDefinition);
```
Looks like the pattern "get generator to then generate a name" is a very common pattern (seen above as well). Does it maybe make sense to — instead of exposing the generator — introduce a shortcut method on the configuration source to generate a bean name for a definition?

olivergierke commented on src/main/java/org/springframework/data/repository/config/RepositoryConfigurationSourceSupport.java:
```diff
@@ -28,26 +28,32 @@
-	public RepositoryConfigurationSourceSupport(Environment environment) {
+	public RepositoryConfigurationSourceSupport(Environment environment, ClassLoader classLoader) {
 
 		Assert.notNull(environment, "Environment must not be null!");
```
I'd rather put a blank line after the assert and remove the one below to group the assignments.

olivergierke commented on src/main/java/org/springframework/data/repository/config/RepositoryConfigurationSourceSupport.java:
```diff
@@ -28,26 +28,32 @@
-	 * 
+	 *
 	 * @param environment must not be {@literal null}.
+	 * @param classLoader
```
Nullability requirements? Missing assertion?

olivergierke commented on src/main/java/org/springframework/data/repository/config/RepositoryConfigurationSourceSupport.java:
```diff
@@ -98,10 +104,19 @@ public RepositoryConfigurationSourceSupport(Environment environment) {
 		return false;
 	}
+
+	/*
```
Formatting seems off here. Caused by spaces maybe?

olivergierke commented on src/main/java/org/springframework/data/repository/config/SelectionSet.java:
```diff
@@ -0,0 +1,101 @@
+ */
+class SelectionSet<T> {
+
+	private Collection<T> collection;
```
Final?

olivergierke commented on src/main/java/org/springframework/data/repository/config/SelectionSet.java:
```diff
@@ -0,0 +1,101 @@
+			return;
+		}
+
+		collection = collection.stream().filter(predicate).collect(Collectors.toList());
```
Shouldn't we rather return a new instance?

olivergierke commented on src/main/java/org/springframework/data/repository/config/SelectionSet.java:
```diff
@@ -0,0 +1,101 @@
+	 *
+	 * @param predicate To be used for filtering.
+	 */
+	void filter(Predicate<T> predicate) {
```
Should this be `filterIfNecessary(…)` instead? With `filter(…)` I don't necessarily expect the predicate not to be applied in case there already is a unique result.

olivergierke commented on src/test/java/org/springframework/data/repository/cdi/CdiRepositoryBeanUnitTests.java:
```diff
@@ -119,6 +124,36 @@ public void createsPassivationId() {
+
+		CustomRepositoryImplementationDetector detector = mock(CustomRepositoryImplementationDetector.class);
+
+		CdiRepositoryBean<SampleRepository> bean = new CdiRepositoryBean<SampleRepository>(
```
`//` needed.

olivergierke commented on src/test/java/org/springframework/data/repository/config/DefaultRepositoryConfigurationUnitTests.java:
```diff
@@ -17,29 +17,44 @@
+	@Before
+	public void before() {
+
+		when(source.getBeanNameGenerator()).thenReturn(
```
Extract variable and remove breaks.

olivergierke commented on src/test/java/org/springframework/data/repository/config/DefaultRepositoryConfigurationUnitTests.java:
```diff
@@ -48,4 +63,15 @@ public void supportsBasicConfiguration() {
 		assertThat(configuration.isLazyInit(), is(false));
 	}
+
+	private RootBeanDefinition createBeanDefinition() {
```
Static?

olivergierke commented on src/test/java/org/springframework/data/repository/config/SelectionSetUnitTests.java:
```diff
@@ -0,0 +1,64 @@
+
+		assertEquals("one", selection.uniqueResult());
+	}
+}
```
Line break.

olivergierke commented on src/main/java/org/springframework/data/repository/config/SelectionSet.java:
```diff
@@ -0,0 +1,101 @@
+ *
+ * @author Jens Schauder
+ */
+class SelectionSet<T> {
```
I am wondering whether we shouldn't let `uniqueResult()` return an `Optional<T>` so that the client is forced to deal with the absence of the a value. Currently in that case we just throw an `IllegalStateException` saying "More then one element in collection." which doesn't give a lot of context. I'd rather let the client — who has that context — be forced to mitigate between the `Optional<T>` and `T` as it know what exactly the absence means and can craft a way more meaningful exception. With the current state, the client basically assumes the call would produce something reasonable and moves on.

olivergierke commented on src/main/java/org/springframework/data/repository/config/SelectionSet.java:
```diff
@@ -0,0 +1,101 @@
+		collection = collection.stream().filter(predicate).collect(Collectors.toList());
+	}
+
+	private static <S> Function<Collection<S>, S> defaultCallback() {
```
Should this be `default*F*allback()`?

olivergierke commented on src/main/java/org/springframework/data/repository/config/SelectionSet.java:
```diff
@@ -0,0 +1,101 @@
+
+	private static <S> Function<Collection<S>, S> defaultCallback() {
+
+		return c -> {
```
Ternary if?

olivergierke commented on src/main/java/org/springframework/data/repository/config/SelectionSet.java:
```diff
@@ -0,0 +1,101 @@
+
+	private T findUniqueResult() {
+
+		if (collection.size() == 1) {
```
Ternary if?

olivergierke commented on src/main/java/org/springframework/data/repository/config/RepositoryConfigurationSource.java:
```diff
@@ -118,4 +120,22 @@
+	 * @return must not be {@literal null}
+	 * @since 2.0
+	 */
+	RepositoryBeanNameGenerator getBeanNameGenerator();
```
We shouldn't need this method anymore with `generateBeanName(…)` introduced, right? If that's the case, I'd rather drop the default implementation and move the code into the implementation base class.

olivergierke commented on src/main/java/org/springframework/data/repository/config/RepositoryConfigurationSourceSupport.java:
```diff
@@ -28,26 +28,33 @@
 	protected static final String DEFAULT_REPOSITORY_IMPL_POSTFIX = "Impl";
 
 	private final Environment environment;
 
```
Remove blank line.

schauder replied:
done

olivergierke commented on src/test/java/org/springframework/data/repository/config/CustomRepositoryImplementationDetectorUnitTests.java:
```diff
@@ -0,0 +1,114 @@
+
+		System.out.println(beanDefinition);
+	}
+}
```
Line break.

schauder replied:
done (and remove the sout as well)

olivergierke commented on src/main/java/org/springframework/data/repository/config/SelectionSet.java:
```diff
@@ -0,0 +1,97 @@
+	 *
+	 * @param predicate To be used for filtering.
+	 */
+	SelectionSet<T> filter(Predicate<T> predicate) {
```
let's rename this to `filterIfNecessary(…)` to make obvious that the given `Predicate` might not be applied.

schauder replied:
ok

xhaggi commented on src/test/java/org/springframework/data/repository/config/SelectionSetUnitTests.java:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	@Test // DATACMNS-764
+	public void useseFallbackWhenConfigured() {
```
typo here?

xhaggi commented on src/main/asciidoc/repositories.adoc:
```diff
@@ -649,6 +649,37 @@ If you use namespace configuration, the repository infrastructure tries to autod
 
+===== Resolution of ambiguity
+
+If multiple implementations with matching class names get found in different packages Spring Data uses the bean names to identify the correct on to use.
```
Typo: correct **one** to use.

xhaggi commented on src/main/asciidoc/repositories.adoc:
```diff
@@ -649,6 +649,37 @@ If you use namespace configuration, the repository infrastructure tries to autod
+
+If multiple implementations with matching class names get found in different packages Spring Data uses the bean names to identify the correct on to use.
+
+Given the following two custom implementations for the `UserRepository` introduced above the first implementation will get picked. It's bean name is `userRepositoryImpl` matches that of the repository interface (`userRepository`) plus the postfix `Impl`.
```
Type: **Its** bean name

xhaggi commented on src/main/java/org/springframework/data/repository/config/SelectionSet.java:
```diff
@@ -0,0 +1,95 @@
+	 * be identified the callback passed in at the constructor gets called and its return value becomes the return value
+	 * of this method.
+	 *
+	 * @return a unique result, or the result of the callback provided in the constructor.
```
The javadoc is a bit confusing. The method does not apply any filter, because there is no filter applied if only one item exists within the collection. And you talk about a callback, but in parameter in the Ctor is named fallback.

schauder commented:
Rebased the branch to incorporate changes on master: 
* Usage of Optional
* Update of Mockito version

Also fixed the issues pointed out by @xhaggi
